### PR TITLE
Use name to build friendly_ids for custom repos

### PR DESCRIPTION
**Pull request by tmuntaner**
## Description

In the demo of the friendly_id feature, the MD5 of URLs was seen as not as user friendly. This is an attempt to use name instead.

## Change Type

*Please select the correct option.*

- [x] **New Feature** (a non-breaking change which adds new functionality)

## Checklist

*Please check off each item if the requirement is met.*

- [x] I have verified that my code follows RMT's coding standards with `rubocop`.
- [x] I have reviewed my own code and believe that it's ready for an external review.
- [x] I have provided comments for any hard-to-understand code.
- [x] I have documented the `MANUAL.md` file with any changes to the user experience.
- [x] RMT's test coverage remains at 100%.
- [x] If my changes are non-trivial, I have added a changelog entry to notify users at `package/obs/rmt-server.changes`.

### CODE REVIEW

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/138923/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged f6503d6a26b202c0da3572e136df88a768cfe068 into 8611a3d5b19d8a3ea3bef671fb5c4b05defad8c1)

**Review by thutterer — COMMENTED** [review]
Thanks for making the effort for an alternative to the MD5 sums! :+1: 

I think this is a step in the right direction, and as most users probably haven't set up too many custom repos with identical names anyway, the whole "append-1" logic won't confuse to many folks.

Yet I think we maybe could do even better. Let me try to explain my original thought when I voted for "use the name":

I thought we could solve this without any "append numbers when needed" as I find this not totally intuitive. Because I as a user would prefer if RMT just makes me use unique names, and doesn't work around the fact that I haven't. I know, so far RMT hasn't enforced unique names, so there could be users that have two custom repos called "foo". Unlikely, but totally possible.

So how about this instead:
Let's make a breaking change, RMT 3.0, where we run a database migration that makes the name column unique* once and for all, and only once runs a bit of custom code, very similar to what you have in this PR, but only as part of the migration, renaming one of my two "foo" repos to "foo-1", warning me about this change in the changelog and migration output.

After the migration, no custom code or "dynamically unique name" is required.

Do you think that would work?

*: In case a db-level uniqueness would clash with non-custom repos not always having unique names, we could instead only use a Rails validation, scoped to custom repos.

**Comment by tmuntaner** [comment]
@thutterer 

Thanks for your feedback, but I don't think this would work the way we want:

* SCC repositories can and will have duplicated `name`s.
* If we keep the non-uniqueness of `name` and only enforce uniqueness on custom repositories in code, we cannot trust the data to always be unique.
* `name` could be unique for custom repositories, but not for SCC repositories, so we would need two separate columns in the database to target each type.

What I want is this: give a good user experience, ensure that RMT always has good data, and lastly, reduce the developer confusion.

Previously not-so-nice issues with RMT:

* We treated `scc_id` liked it was unique in our code, but it wasn't and we're lucky it didn't cause any major issues. The only unique index we had was for `external_url`. See my recently discovered bug https://github.com/SUSE/rmt/pull/641 which this caused as a side effect.
* SCC and custom repositories used separate database columns for their find methods. This caused developer confusion and the divergence leaked to the UI and reduced user friendliness.

I fear that this suggestion would bring us back to the previous state and I don't see the gained benefit outweighing it, yet.

This was the best way I found so far, but I'm still open to further improvements.

**Comment by thutterer** [comment]
> `name` could be unique for custom repositories, but not for SCC repositories, so we would need two separate columns in the database to target each type.

That's not necessarily true. As I mentioned at the end of my previous column, we could do this in Railsland:

```ruby
validates :name, uniqueness: true, if: :custom_repo?
```

**Comment by tmuntaner** [comment]
@thutterer yes, we could be reasonably confident that `name` is unique for custom repos (not fully confident as the database isn't also keeping this unique), but we still would need two find columns for custom and scc repos.

**Comment by tmuntaner** [comment]
@lagartoflojo Here is the attempt for a human readable ID if you would like to leave thoughts.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/139353/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged f6503d6a26b202c0da3572e136df88a768cfe068 into b6fafd48b69278d3ec13deec72d48082e67346a4)

**Comment by lagartoflojo** [comment]
Thanks for the pointer, @tmuntaner.
Could you provide a sample workflow (using rmt-cli commands) in the PR description? Then it's more clear what this PR achieves from the user perspective. I (as PO) don't want to dig into the implementation details =)

**Comment by tmuntaner** [comment]
@lagartoflojo here is a workflow I just ran.

```
rmt-cli repos custom add https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.1/ dummy-repo
rmt-cli repos custom add https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.2/ dummy-repo
rmt-cli repos custom add https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.3/ dummy-repo
rmt-cli repos custom add https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.4/ dummy-repo
rmt-cli repos custom add https://foo.bar.com/ foo-bar
rmt-cli repos custom add https://bar.foo.com/ "boo foo"
```

```
rmt-cli repos custom ls
```

```
+--------------+------------+-----------------------------------------------------------------------------------------------+---------------+---------+---------------+
| ID           | Name       | URL                                                                                           | Mandatory?    | Mirror? | Last Mirrored |
+--------------+------------+-----------------------------------------------------------------------------------------------+---------------+---------+---------------+
| boo-foo      | boo foo    | https://bar.foo.com/                                                                          | Not Mandatory | Mirror  |               |
| dummy-repo   | dummy-repo | https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.1/ | Not Mandatory | Mirror  |               |
| dummy-repo-1 | dummy-repo | https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.2/ | Not Mandatory | Mirror  |               |
| dummy-repo-2 | dummy-repo | https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.3/ | Not Mandatory | Mirror  |               |
| dummy-repo-3 | dummy-repo | https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.4/ | Not Mandatory | Mirror  |               |
| foo-bar      | foo-bar    | https://foo.bar.com/                                                                          | Not Mandatory | Mirror  |               |
+--------------+------------+-----------------------------------------------------------------------------------------------+---------------+---------+---------------+
```

**Comment by lagartoflojo** [comment]
Couple of questions come to mind:

- How does RMT determine what is "bad input"? Why is it bad?
- How is the repo name used by zypper when enabling a repo on a client machine (if at all)?

**Comment by lagartoflojo** [comment]
Other than that, I think this looks great! It's probably a very low chance that people are really passing `"quoted strings"` as the repo name.
What happens if you don't quote spaces?
e.g. `rmt-cli repos custom add https://foo.bar.com/ foo bar`

**Comment by tmuntaner** [comment]
@lagartoflojo 

When making the `friendly_id`, I made it so it stripped out any character that isn't a `-`, `_`, `a-Z`, or `0-9`. Accepting anything else makes using it as user input harder because characters like `$`. They just don't play friendly with the terminal and copy pasta.

https://github.com/SUSE/rmt/pull/640/files#diff-ad849f0f01f3789c134cc112da76dc37R38

---

From what I could find, `zypper` uses the name as a readable identifier. It's actually quite liberal in accepting any type of name.

```
docker run --rm -it opensuse/leap:15.1 /bin/bash
zypper addrepo https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.1/ "foo bar"
zypper addrepo https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.1/ "foo bar \$ asd "
zypper addrepo https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.1/ foobar
```

```
zypper lr -u
```

```
Repository priorities are without effect. All enabled repositories share the same priority.

#  | Alias                     | Name                               | Enabled | GPG Check | Refresh | URI                                                                                          
---+---------------------------+------------------------------------+---------+-----------+---------+----------------------------------------------------------------------------------------------
 1 | foo bar                   | foo bar                            | Yes     | (r ) Yes  | No      | https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.1/
 2 | foo bar $ asd             | foo bar $ asd                      | Yes     | (r ) Yes  | No      | https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.1/
 3 | foobar                    | foobar                             | Yes     | (r ) Yes  | No      | https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.1/
 4 | repo-debug                | Debug Repository                   | No      | ----      | ----    | http://download.opensuse.org/debug/distribution/leap/15.1/repo/oss/                          
 5 | repo-debug-non-oss        | Debug Repository (Non-OSS)         | No      | ----      | ----    | http://download.opensuse.org/debug/distribution/leap/15.1/repo/non-oss/                      
 6 | repo-debug-update         | Update Repository (Debug)          | No      | ----      | ----    | http://download.opensuse.org/debug/update/leap/15.1/oss/                                     
 7 | repo-debug-update-non-oss | Update Repository (Debug, Non-OSS) | No      | ----      | ----    | http://download.opensuse.org/debug/update/leap/15.1/non-oss/                                 
 8 | repo-non-oss              | Non-OSS Repository                 | Yes     | (r ) Yes  | No      | http://download.opensuse.org/distribution/leap/15.1/repo/non-oss/                            
 9 | repo-oss                  | Main Repository                    | Yes     | (r ) Yes  | No      | http://download.opensuse.org/distribution/leap/15.1/repo/oss/                                
10 | repo-source               | Source Repository                  | No      | ----      | ----    | http://download.opensuse.org/source/distribution/leap/15.1/repo/oss/                         
11 | repo-source-non-oss       | Source Repository (Non-OSS)        | No      | ----      | ----    | http://download.opensuse.org/source/distribution/leap/15.1/repo/non-oss/                     
12 | repo-update               | Main Update Repository             | Yes     | (r ) Yes  | No      | http://download.opensuse.org/update/leap/15.1/oss/                                           
13 | repo-update-non-oss       | Update Repository (Non-Oss)        | Yes     | (r ) Yes  | No      | http://download.opensuse.org/update/leap/15.1/non-oss/
```

---

```
rmt-cli repos custom add https://bar.foo.com/ boo foo
```

```
ERROR: "rmt-cli custom add" was called with arguments ["https://bar.foo.com/", "boo", "foo"]
Usage: "rmt-cli repos custom add URL NAME"
```

zypper does the same

```
zypper addrepo https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.1/ foo bar
```

```
Too many arguments.
Usage:
addrepo (ar) [OPTIONS] <URI> <ALIAS>
addrepo (ar) [OPTIONS] <FILE.repo>

Add a repository to the system. The repository can be specified by its URI or can be read from
specified .repo file (even remote).

  Command options:

-r, --repo <FILE.repo>      Just another means to specify a .repo file to read.
-c, --check                 Probe URI.
-C, --no-check              Don't probe URI, probe later during refresh.
-p, --priority <PRIORITY>   Set priority of the repository. Default: 0
-k, --keep-packages         Enable RPM files caching. Default: false
-K, --no-keep-packages      Disable RPM files caching.
-g, --gpgcheck              Enable GPG check for this repository.
    --gpgcheck-strict       Enable strict GPG check for this repository.
    --gpgcheck-allow-unsigned
                            Short hand for '--gpgcheck-allow-unsigned-repo
                            --gpgcheck-allow-unsigned-package'.
    --gpgcheck-allow-unsigned-repo
                            Enable GPG check but allow the repository metadata to be unsigned.
    --gpgcheck-allow-unsigned-package
                            Enable GPG check but allow installing unsigned packages from this
                            repository.
-G, --no-gpgcheck           Disable GPG check for this repository.
    --default-gpgcheck      Use the global GPG check setting defined in /etc/zypp/zypp.conf. This is
                            the default.
-n, --name <NAME>           Set a descriptive name for the repository.
-e, --enable                Enable a disabled repository. Default: true
-d, --disable               Disable the repository (but don't remove it).
-f, --refresh               Enable auto-refresh of the repository. Default: false
-F, --no-refresh            Disable auto-refresh of the repository.

  Legacy options:

-t, --type                  The repository type is always autodetected. This option is ignored.
```

**Comment by lagartoflojo** [comment]
> When making the friendly_id, I made it so it stripped out any character that isn't a -, _, a-Z, or 0-9. 

What about non-English characters, like á, or ñ? =)

> Accepting anything else makes using it as user input harder because characters like $. They just don't play friendly with the terminal and copy pasta.

The `cp` command lets one use whatever characters one wants to name files. The onus is on the user to properly use their terminal. I'm not sure why RMT should behave differently (since RMT itself doesn't care if there are funny symbols in the repo name).

Anyway, I think using weird characters to name your repo is unlikely anyway. So I don't think we need to go deeper in the discussion.

Overall it feels like using repo name as the identifier for running subsequent commands gives a much nicer user experience 👍

**Comment by tmuntaner** [comment]
@lagartoflojo you were right, I was a bit English-language normative there and forgot about other languages ^_^

It's fixed it in my latest commit. For the sanitation of IDs, I feel that it will be easier to loosen our strictness than the other way around, Maybe we can wait for some user feedback to see if this causes issues for people.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/139357/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged f6503d6a26b202c0da3572e136df88a768cfe068 into b7f895aae2d12863d7317680e9f627b955f37efd)

**Comment by suse-tests-pass** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>spec/models/repository_spec.rb</td>
      <td>156</td>
      <td>Add an empty line after the last <code>let</code> block. (https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/EmptyLineAfterFinalLet)</td>
    </tr>
  </tbody>
</table></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/139358/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged a66d61ab35cfd579979518580f184c9be91621de into b7f895aae2d12863d7317680e9f627b955f37efd)

File: app/models/repository.rb [hunk 1]
```diff
@@ -34,6 +34,22 @@ def make_local_path(url)
       (path == '') ? '/' : path
     end
 
+    def make_friendly_id(input)
```
Comment (Sergeykot) [inline app/models/repository.rb:37]: maybe `generate_friendly_id` ?

**Review by Sergeykot — APPROVED** [review]
LGTM :+1:

**Review by lcaparroz — APPROVED** [review]
LGTM :+1:

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/140254/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged abd74e6eb21cd0bc53a6cb0099073b34be88511f into b7f895aae2d12863d7317680e9f627b955f37efd)

**Review by lcaparroz — APPROVED** [review]
Nice! The update worked :)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/140256/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 0d40766cf4bc9017d231712aac33e76ad9e84f4c into b7f895aae2d12863d7317680e9f627b955f37efd)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/140402/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-run-tests. (Merged 083ca33dbfd25432216cc2314632b5311994b149 into b7f895aae2d12863d7317680e9f627b955f37efd)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/140403/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged b68d40f071be19b61b8b4d8ce13903fcc0ca9817 into b7f895aae2d12863d7317680e9f627b955f37efd)

**Review by lcaparroz — APPROVED** [review]
New changes worked fine when I tested :+1:

**Comment by tmuntaner** [comment]
Thanks @Sergeykot @lcaparroz

**Pull request merged**